http-common/src/connector.rs

```diff
@@ -582,22 +672,85 @@ fn get_systemd_socket() -> Result<Option<std::os::unix::io::RawFd>, String> {
-            Err(err @ std::env::VarError::NotUnicode(_)) => {
-                return Err(format!("could not read LISTEN_FDS env var: {}", err))
-            }
+fn socket_name_to_fd(name: &str) -> Result<std::os::unix::io::RawFd, String> {
+    let listen_fdnames = match get_env("LISTEN_FDNAMES")? {
+        Some(listen_fdnames) => listen_fdnames,
+        None => return Err("LISTEN_FDNAMES not found".to_string()),
+    };
+
+    let listen_fdnames: Vec<&str> = listen_fdnames.split(':').collect();
+
+    let index: std::os::unix::io::RawFd =
+        match listen_fdnames.iter().position(|&fdname| fdname == name) {
+            Some(index) => match std::convert::TryInto::try_into(index) {
+                Ok(index) => index,
+                Err(_) => return Err("couldn't convert LISTEN_FDNAMES index to fd".to_string()),
+            },
+            None => return Err(format!("socket {} not found", name)),
         };
-        let listen_fds: std::os::unix::io::RawFd = listen_fds
+
+    // The index in LISTEN_FDNAMES is an offset from SD_LISTEN_FDS_START.
+    Ok(index + SD_LISTEN_FDS_START)
+}
+
+#[cfg(feature = "tokio1")]
+fn fd_to_listener(fd: std::os::unix::io::RawFd) -> std::io::Result<Incoming> {
+    if is_unix_fd(fd)? {
+        let listener: std::os::unix::net::UnixListener =
+            unsafe { std::os::unix::io::FromRawFd::from_raw_fd(fd) };
+        listener.set_nonblocking(true)?;
+        let listener = tokio::net::UnixListener::from_std(listener)?;
+        Ok(Incoming::Unix {
+            listener,
+            user_state: Default::default(),
+        })
+    } else {
+        let listener: std::net::TcpListener =
+            unsafe { std::os::unix::io::FromRawFd::from_raw_fd(fd) };
+        listener.set_nonblocking(true)?;
+        let listener = tokio::net::TcpListener::from_std(listener)?;
+        Ok(Incoming::Tcp { listener })
+    }
+}
+
+/// Finds the number of available systemd sockets. Checks if this process has been socket-activated.
+///
+/// This mimics `sd_listen_fds` from libsystemd, then returns the number of systemd socket fds.
+#[cfg(feature = "tokio1")]
+fn get_systemd_sockets() -> Result<Option<std::os::unix::io::RawFd>, String> {
+    // Ref: <https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html>
+    //
+    // >sd_listen_fds parses the number passed in the $LISTEN_FDS environment variable, then sets the FD_CLOEXEC flag
+    // >for the parsed number of file descriptors starting from SD_LISTEN_FDS_START. Finally, it returns the parsed number.
+    //
+    // Note that this function always returns the first fd. It cannot be used for processes which expect more than one socket.
+    // CS/IS/KS only expect one socket, so this is fine, but it is not the case for iotedged (mgmt and workload sockets) for example.
+    //
+    // If obtaining more than one fd is required in the future, keep in mind that it requires getting fds by name (by inspecting the LISTEN_FDNAMES env var)
```

Comment:
The future is now, so update this comment.

Reply:
Updated